packages/apollo-server-codegen/ARCHITECTURE.md

```diff
@@ -0,0 +1,67 @@
+# Architecture
+
+## Overview
+
+The code generator is a set of modules designed to accept a GraphQL schema, and from it generate type definitions in TypeScript format (other languages may be supported in the future).
+
+Similar to projects like [graphql code generator](https://graphql-code-generator.com), the generated type definitions describe both the type of the GraphQL objects represented in the schema, and the type of the `resolvers` function that an Apollo Server serving would require to serve the schema.
+
+However, this project expands upon existing efforts by providing accurate type definitions for federated schemas, built using the recently announced Apollo Federation platform.
+
+This package serves only as an API, to interface with it we suggest either the `Apollo CLI`, where it is available as `apollo server:codegen`, or the `[.....]` plugin we have built for `graphql-code-generator`.
+
+## Design
+
+The generator is built using a two-layered approach, where the frontend is able to convert an SDL-formatted input string to an internal representation (`IR`), and one or backends are able to convert that `IR` to files of type declaration source code in a particular programming language. Currently only a backend for TypeScript is provided.
+
+The two layers interact via a [visitor pattern](https://en.wikipedia.org/wiki/Visitor_pattern) interface, which allows new backends to be created without any code change to the frontend. To add a new language backend, one must implement the [Translator](Translators/index.ts) interface for that language, which is a declaration of how each `IR` node (things like `IR.EnumDefinition`, `IR.UnionDefinition`, etc.), should be represented in a given target language. For example, the TypeScript implementation implements the `Translator` interface for `IR.Description` like so:
+
+```ts
+  public translateDescription(t: IR.Description): string {
+    return [
+      "/**",
+      ...t.description.split("\n").map(line => `* ${line}`),
+      " */"
+    ].join("\n");
+  }
+```
+
+Additionally, every node in the `IR` implements [`Translateable`](Translators/index.ts):
+
+```ts
+export interface Translatable {
+  translate(t: Translator): string | undefined;
+}
+```
+
+This allows a `Translator`'s translation to recursively descend into the `IR` tree by simply calling `.translate(this)` on each IR node, and the node will handle delegation back to the `Translator`.
```

Comment:
Can we rephrase? I don't follow what this means:
> and the node will handle delegation back to the `Translator`

Reply:
Yeah I think its fine to just remove that part.